[Server] DB 환경 설정

soomanbaek opened the pull request:
## 🤷‍♂️ Description
- #8 

MySQL과 mongoDB를 Prisma와 연결
[연결하는 과정](https://black-tokkies.notion.site/Prisma-DB-5e63a921b84749549438e11f0bcf8798)
## 📝 Primary Commits
- MySQL 스키마 생성 (User, Panel, Question, Answer)
- mongoDB 스키마 생성 (ToquizUser)
- MysqlPrismaService, MongodbPrismaService 생성

## 📷 Screenshots
<img width="366" alt="스크린샷 2023-02-01 오후 11 37 24" src="https://user-images.githubusercontent.com/72093196/216072633-0ed4205d-6eb0-450f-b79b-6dc68a84c8cc.png">

soomanbaek committed:
chore: .gitignore 수정 (#1)

soomanbaek committed:
feat: prisma 설치 및 mysql Schema 생성 (#8)

soomanbaek committed:
feat: prisma service 생성 및 panel, user 모듈에 import (#8)

soomanbaek committed:
chore: Eslint, Prettier 수정 (#1)

soomanbaek committed:
build: package.json 수정 (#8)

- `-`에서`:`으로 컨벤션 통일
- mysql prisma migrate 스크립트 추가

soomanbaek committed:
feat: 컨벤션 통일로 인한 Answer 스키마 수정 (#8)

soomanbaek committed:
feat: mysql schema 네이밍 컨벤션 수정 (#8)

- model field `snake_case` -> `camelCase`

soomanbaek committed:
feat: mysql schema 파일 명, package.json script 수정 (#8)

- `schema.prisma` -> `mysql.schema.prisma`
- mysql migrate script 수정

soomanbaek committed:
feat: mongodb schema 생성 (#8)

soomanbaek committed:
feat: mongodb schema 수정 (#8)

- output 옵션 추가 [DB 2개 사용을 위해]

soomanbaek committed:
feat: prisma.service 생성 (#8)

soomanbaek committed:
feat: 생성한 prismaService 모듈에 추가 (#8)

soomanbaek committed:
build: package.json script 수정 (#8)

soomanbaek committed:
feat: mongodb schema 수정 (#8)

soomanbaek committed:
feat: prismaService 수정 (#8)

soomanbaek committed:
build: package.json script 수정 (#8)

- package.json mongodb:migrate:dev 삭제 (mongodb는 현재 prisma에서 migrate 지원 안함)

soomanbaek committed:
build: yarn.lock 수정 (#8)

leegwae commented on server/package.json:
```diff
@@ -13,21 +13,27 @@
     "@nestjs/core": "^9.0.0",
     "@nestjs/platform-express": "^9.0.0",
+    "@prisma/client": "^4.9.0",
+    "dotenv-cli": "^7.0.0",
```
`dotenv-cli`는 어떤 기능을 할 수 있는 라이브러리인가요?

leegwae commented on server/prisma/mongodb.schema.prisma:
```diff
@@ -0,0 +1,25 @@
+  questions String[]
+}
+
+model ToquizUser {
+  id        String   @id @default(auto()) @map("_id") @db.ObjectId
+  pannel    Pannel[]
+  createdAt DateTime @default(now())
+  updatedAt DateTime @updatedAt
```
`User`와 `ToquizUser`는 이름만 놓고 보면 구별하기 어려워보입니다. 비회원이나 익명 사용자 같은 건 어떠신가요?

leegwae replied:
```typescript
pannel Pannel[]
```
이 필드의 이름도 `pannels`처럼 복수형으로 맞춰주는게 통일된 거 같네요 ㅎㅎ

soomanbaek replied:
오 좋은거 같아요! 수정하겠습니다 👍

leegwae commented on server/prisma/mysql.schema.prisma:
```diff
@@ -0,0 +1,88 @@
+  deletedAt DateTime?
+}
+
+model Scrap {
+  user     User   @relation(fields: [userId], references: [id])
+  userId   String
+  pannel   Pannel @relation(fields: [pannelId], references: [id])
+  pannelId Int
+
+  createdAt DateTime  @default(now())
+  updatedAt DateTime  @updatedAt()
+  deletedAt DateTime?
+
+  @@id([userId, pannelId])
+}
```
스크랩은 토글 같은 거라고 생각했는데 얘는 스크랩하고 스크랩 취소할 때 C/U/D가 어떻게 되나요?

soomanbaek replied:
스크랩은 저도 토클이라고 생각하고 있어요!
`Create`가 될 때는 패널의 스크랩을 누르면 Scrap 테이블에 userId와 pannelId가 생성되고,
`Update`는 딱히 없을거 같고,
`Delete`를 하게되면 Scrap 테이블에서 hard delete하면 될꺼 같네요.

생각해보니 updatedAt과 deletedAt은 없어도 될꺼 같네요!

leegwae commented on server/prisma/mysql.schema.prisma:
```diff
@@ -0,0 +1,88 @@
+  @@id([userId, pannelId])
+}
+
+model Pannel {
+  id Int @id @default(autoincrement())
+
+  user   User   @relation(fields: [userId], references: [id])
+  userId String
+
+  scrappedUsers Scrap[]
+  questions     Question[]
+
+  describe   String
```
너무 늦게 발견한 거 같긴 한데 `describe`은 동사니까 description 같이 명사를 쓰는게 좋지 않을까요 ㅋㅋㅋ 혹시 길어서 줄이신 건가요?

soomanbaek replied:
그러네요 ㅋㅋㅋ 길어서 줄인건 아니고 캐치하지 못했었네요..

leegwae commented on server/prisma/mongodb.schema.prisma:
```diff
@@ -0,0 +1,25 @@
+  url      = env("MONGODB_DATABASE_URL")
+}
+
+type Pannel {
+  pannelId  Int
+  likes     String[]
+  questions String[]
+}
```
`likes`와 `questions`에는 MySQL에 저장된 질문의 id가 들어가는 거죠?

soomanbaek replied:
네네 맞아요!

leegwae commented on server/apps/panel/src/panel.module.ts:
```diff
@@ -1,10 +1,11 @@
   imports: [],
   controllers: [PanelController],
-  providers: [PanelService],
+  providers: [PanelService, MysqlPrismaService, MongodbPrismaService],
```
`PanelService`가 `**PrismaService`를 가져다 쓰게 되는건가요? 아니면 그 위의 레이어인 Controller에서 Service를 조합해서 사용하는 방식이 되나요?

soomanbaek replied:
지금 `Repository 패턴`을 사용하는 것도 고려하고 있는데, 사용하게 되면  `PanelRepository`에서 `**PrismaService`를 가져다 쓰게 돼요. 그렇지 않으면 `PanelService`가 `**PrismaService`를 가져다 쓰게 구현 하려고요!

leegwae reviewed (approved):
고생 많으셨습니다^^~

soomanbaek committed:
fix: #12 코드리뷰에 따른 수정 (#8)

- `pannel` -> `panel`
- `ToquizUser` -> `AnonymousUser`
- `scrappedPannels`, `scrappedUsers` -> `scraps`
- `describe` -> `description`
